CiviImport - Allow multiple dedupe rules to be selected

colemanw opened the pull request:
Overview
----------------------------------------
Alters the CIviImport UI to allow multiple dedupe rules. Especially useful with the Contact Type is unspecified:

<img width="1910" height="556" alt="image" src="https://github.com/user-attachments/assets/e8201144-8c0e-4ba8-ae8f-987534196c8f" />



Comments
----------------------------------------
@eileenmcnaughton - updated PR to add icons and optgroups.

colemanw committed:
CiviImport - Allow multiple dedupe rules to be selected

colemanw commented:
@eileenmcnaughton yea I could add optgroups. Looks like there are also some generic ones that work for any contact type?

eileenmcnaughton commented:
@colemanw yeah  - the ID / ext identifier is ALWAYS the fall back - so it only makes sense if nothing else is used. Also the Unique email is a special 'cross-type' rule

eileenmcnaughton commented:
@colemanw cool - now it just needs to work on the back end to 'detect' the type of contact to select the rule

colemanw committed:
CiviImport - Add guessContactType function

colemanw committed:
CiviImport - Use dedupe rules according to contact type

Whether explicitly set or inferred, this selects the appropriate rule to use for each row.

eileenmcnaughton commented:
The first test error

CiviApiImportTest::testApiActions
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'No matching Contact found'
+'Expected to find one DedupeRuleGroup record, but there were zero.'

/home/homer/buildkit/build/build-2/web/sites/all/modules/civicrm/ext/civiimport/tests/phpunit/CiviApiImportTest.php:142
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2538

eileenmcnaughton commented on CRM/Import/Parser.php:
```diff
@@ -1484,4 +1468,25 @@ protected function removeEmptyValues($array) {
+   * @param array $values
+   * @return string
+   */
+  protected function guessContactType(array $values): string {
```
I think this can got in `ImportParser` (which is part of civiimport & not shared with the contact import)

eileenmcnaughton commented:
Opps on import

<img width="1351" height="471" alt="image" src="https://github.com/user-attachments/assets/02b00f54-187a-4b78-a7d5-669a51c6eb41" />

eileenmcnaughton commented:
@colemanw  just thinking through what a test would look like - lets say 

- Dedupe rules 
 1) Individual - email + first name + last name
 2) Organization email + organization name
 3) Individual email
 
Rows
1) first + last + email matches an existing individual
2) same email but organization name & matches an existing org
3) same email (matches an an existing individual who has a different name )
4) first + last + email - different email - no existing matches - should create org
5) email + org name - no existing matches should create org
6) email only - no matches - should create individual

colemanw commented on tests/phpunit/CRM/Contribute/Import/Parser/ContributionTest.php:
```diff
@@ -347,12 +347,12 @@ public function doUserJobImport(array $importMappings): array {
         'action' => 'create',
         'contact_type' => 'Organization',
-        'dedupe_rule' => 'OrganizationUnsupervised',
+        'dedupe_rule' => ['OrganizationUnsupervised'],
```
@eileenmcnaughton my preferred approach is usually to do the conversion at read-time rather than as an upgrade step. That way as the UserJob is being read for any purpose (editing it, or using it) it gets converted to the right format, and then will be compatible with the UI changes and will get saved correctly.

eileenmcnaughton committed:
Fix test to specific dedupe rule

eileenmcnaughton commented:
Oh dang I didn't push it in - capitalization of branch https://github.com/civicrm/civicrm-core/commit/748b3c571bb650fd622a753b1c7af136b48998dd

eileenmcnaughton committed:
Increase output

eileenmcnaughton committed:
Use Unsupervised rule as default in tests

eileenmcnaughton commented:
@colemanw there are still several places expecting only one dedupe rule - notably in the `MapField` validate functions - which check that sufficient fields have been mapped to meet the dedupe rule requirements (this would ideally be moved to angular layer but....)
```
        $rule = $parser->getDedupeRule('Individual', $self->getUserJob()['metadata']['entity_configuration']['TargetContact']['dedupe_rule'] ?? NULL);
        $errors = $self->validateContactFields($rule, $fields['mapper'], ['external_identifier', 'id']);
```

eileenmcnaughton commented:
ie 

```
CRM_Custom_Import_Parser_ApiTest::testImport
TypeError: Civi\Import\ImportParser::getDedupeRule(): Argument #2 ($name) must be of type ?string, array given, called in /home/homer/buildkit/build/build-4/web/core/ext/civiimport/CRM/CiviImport/Form/MapField.php on line 219

/home/homer/buildkit/build/build-4/web/core/ext/civiimport/Civi/Import/ImportParser.php:297
/home/homer/buildkit/build/build-4/web/core/ext/civiimport/CRM/CiviImport/Form/MapField.php:219
/home/homer/buildkit/build/build-4/web/core/CRM/Custom/Import/Form/MapField.php:43
/home/homer/buildkit/build/build-4/web/core/packages/HTML/QuickForm.php:1600
/home/homer/buildkit/build/build-4/web/core/CRM/Core/Form.php:728
/home/homer/buildkit/build/build-4/web/core/tests/phpunit/CRMTraits/Import/ParserTrait.php:83
/home/homer/buildkit/build/build-4/web/core/tests/phpunit/CRM/Custom/Import/Parser/ApiTest.php:31
/home/homer/buildkit/build/build-4/web/core/tests/phpunit/CiviTest/CiviUnitTestCase.php:4027
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2538
```

colemanw committed:
CiviImport - Fix validation of multiple rules

Also fix looking up '.name' dedupe rules

eileenmcnaughton commented:
Steps to manually test

1) update contact number 1 to have the email jenny@example.com
2) create a new contact first_name = Jen last_name = Lee email = jenny@example.com
3) Contribution import on attached file
[new.csv](https://github.com/user-attachments/files/21725865/new.csv)
with config

<img width="1750" height="961" alt="image" src="https://github.com/user-attachments/assets/5f535592-8b8b-48ad-85e4-283a0ff5f293" />


- expect the first 3 lines to import existing 
Jenny Lee
Default Organization
Jen Lee

- expect the last 3 lines to create new contacts
- Individual Susan Lee
- Organization New Org
- Individual sandy@example.org

colemanw commented on ext/civiimport/CRM/CiviImport/Form/MapField.php:
```diff
@@ -215,8 +215,7 @@ protected function validateRequiredContactFields(string $entity = 'Contact'): ar
     }
-    $parser = $this->getParser();
-    $rule = $parser->getDedupeRule($this->getContactType(), $this->getUserJob()['metadata']['entity_configuration'][$entity]['dedupe_rule'] ?? NULL);
+    $rule = $this->getUserJob()['metadata']['entity_configuration'][$entity]['dedupe_rule'] ?? NULL;
```
@eileenmcnaughton yes the UI puts in an array. What do you mean by "newly created rule?"

eileenmcnaughton replied:
@colemanw sorry an upgraded vs new import not rule - this was just on the demo site for this PR

eileenmcnaughton commented:
CRM_Contribute_Import_Parser_ContributionTest::testImportFieldsValidateCampaignID
Array to string conversion

/home/homer/buildkit/build/build-2/web/core/ext/civiimport/CRM/CiviImport/Form/MapField.php:192
/home/homer/buildkit/build/build-2/web/core/packages/HTML/QuickForm.php:1600
/home/homer/buildkit/build/build-2/web/core/CRM/Core/Form.php:728
/home/homer/buildkit/build/build-2/web/core/tests/phpunit/CRM/Contribute/Import/Parser/ContributionTest.php:785
/home/homer/buildkit/build/build-2/web/core/tests/phpunit/CiviTest/CiviUnitTestCase.php:4027
/home/homer/buildkit/extern/phpunit9/phpunit9.phar:2538

colemanw committed:
CiviImport - Backward-compat for single-valued dedupe_rule (now expects an array)

Also adds backward-compat for rule id provided instead of name

colemanw committed:
CiviImport - Test fixes

eileenmcnaughton committed:
Further test fixes
